File: datastructure/Trie.java
package com.company;

public class Trie {
    public class TrieNode {
        public char val;
        public int count;
        public int endsHere;
        public TrieNode[] children;
        TrieNode(char v) {
            this.val = v;
            this.children = new TrieNode[26];
            for (int i=0;i<26;i++) this.children[i] = null;
        }
    }

    public TrieNode root;

    /** Initialize your data structure here. */
    public Trie() {
        this.root = new TrieNode('/');
    }

    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode cur = this.root;
        int index = -1;
        for (int i=0;i<word.length();i++){
            index = word.charAt(i) - 'a';
            if(cur.children[index] == null) {
                cur.children[index] = new TrieNode(word.charAt(i));
            }
            cur.children[index].count++;
            cur = cur.children[index];
        }
        cur.endsHere++;
    }

    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        TrieNode cur = this.root;
        int index = 0;
        for (int i=0;i<word.length();i++){
            index = word.charAt(i) - 'a';
            if(cur.children[index] == null) return false;

            cur = cur.children[index];
        }
        return cur.endsHere > 0;
    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        TrieNode cur = this.root;
        int index = 0;
        for (int i=0;i<prefix.length();i++){
            index = prefix.charAt(i) - 'a';
            if(cur.children[index] == null) return false;

            cur = cur.children[index];
        }
        return cur. count > 0;
    }
}

File: datastructure/trie.cpp
#include <bits/stdc++.h>

using namespace std;

class Trie {
private:
	vector<Trie*> children;
	int node_count;
	bool leaf;

public:

	Trie(): children(26), node_count(0), leaf(false){
	};


	void insert(string word) {
		Trie* cur = this;
		for(int i = 0; i < word.size(); i++) {
		  int c = word[i] - 'a';
		  if(cur->children[c] == nullptr) {
			cur->children[c] = new Trie();
		  }
		  cur->node_count++;
		  cur = cur->children[c];
		}
		cur->leaf = true;
	}

	Trie* traverse(string word) {
		Trie* cur = this;
		for (int i = 0; i < word.size(); i++){
			int c = word[i] - 'a';
			if(cur->children[c] == nullptr)
				return nullptr;

			cur = cur->children[c];
		}
		return cur;
	}

	bool search(string word) {
		Trie* cur = this->traverse(word);
		return cur == nullptr ? false : cur->leaf;
	}

    bool starts_with(string prefix) {
        Trie* cur = this->traverse(prefix);
		return cur == nullptr ? false : cur->node_count > 0;
    }
};

int main(){
	Trie* trie = new Trie();
	trie->insert("hello");
	trie->insert("world");
	trie->insert("cplusplus");
	cout<< (trie->search("hello") ? "yes" : "no") << endl;
	cout<< (trie->search("world") ? "yes" : "no") << endl;
	cout<< (trie->search("worl") ? "yes" : "no") << endl;
	cout<< (trie->starts_with("worl") ? "yes" : "no") << endl;
	cout<< (trie->starts_with("cplusp") ? "yes" : "no") << endl;
    cout<< (trie->starts_with("cplusps") ? "yes" : "no") << endl;
	return 0;
}
